File: section33#Contours/Python/contours.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# author：lowkeyway time:9/25/2019

import sys
import cv2 as cv
import numpy as np
import matplotlib.pyplot as plt
import random

def contoursTest(imgOri):
    cv.imshow("imgOri", imgOri)
    imgGary = cv.cvtColor(imgOri, cv.COLOR_BGR2GRAY)
    cv.imshow("imgGary", imgGary)
    imgBin = cv.Canny(imgOri, 80, 160)
    cv.imshow("imgBin", imgBin)

    out, contours, hierarchy = cv.findContours(imgBin, cv.RETR_LIST, cv.CHAIN_APPROX_SIMPLE)
    result = np.zeros(imgOri.shape, dtype=imgOri.dtype)
    for index, con in enumerate(contours):
        color = (random.randint(0, 255), random.randint(0, 255), random.randint(0, 255))
        cv.drawContours(imgOri, contours, index, color, 1, 8)
        cv.drawContours(result, contours, index, color, 1, 8)

    cv.imshow("imgMix", imgOri)
    cv.imshow("result", result)

def main_func(argv):
    imgOri = cv.imread("synaptics.jfif")
    if imgOri is None:
        print("Cannot load this picture!")
        return

    contoursTest(imgOri)

    cv.waitKey(0)

if __name__ == '__main__':
    main_func(sys.argv)

File: section33#Contours/C++/contours.cpp
#include <iostream>
#include <string>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

static bool contoursTest(Mat imgOri);
RNG rng(12345);

int main()
{
	Mat imgOri = imread("gril.jfif");
	if (imgOri.empty())
	{
		cout << "Cannot load this picture!" << endl;
		return false;
	}

	contoursTest(imgOri);

	waitKey(0);
	return true;
}

static bool contoursTest(Mat imgOri)
{
	imshow("imgOri", imgOri);

	// binary
	Mat imgGray, imgBin;
	cvtColor(imgOri, imgGray, COLOR_BGR2GRAY);
	GaussianBlur(imgGray, imgGray, Size(5, 5), 0, 0);
	//Canny(imgGray, imgBin, 80, 160);
	threshold(imgGray, imgBin, 0, 255, THRESH_BINARY | THRESH_OTSU);

	//Mat kernel = getStructuringElement(MORPH_RECT, Size(5, 5), Point(-1, -1));
	//morphologyEx(imgBin, imgBin, MORPH_CLOSE, kernel);

	imshow("binary", imgBin);


	vector<vector<Point>> contours;
	vector<Vec4i> hierarchy;
	findContours(imgBin, contours, hierarchy, RETR_TREE, CHAIN_APPROX_SIMPLE);

	Mat result = Mat::zeros(imgOri.size(), imgOri.type());
	Mat pts;
	for (int i = 0; i < contours.size(); i++)
	{
		Scalar color = Scalar(rng.uniform(1, 255), rng.uniform(0, 255), rng.uniform(0, 255));
		drawContours(imgOri, contours, i, color, 1, 8);
		drawContours(result, contours, i, color, 1, 8);
	}

	imshow("imgOri", imgOri);
	imshow("result", result);

	return true;
}